[CS-3762] Firebase Remote Config setup

oidoug opened the pull request:
### Description

This PR adds Firebase's Remote Config packages and Analytics (without Ad ID tracking) and implementation for fetching a few config properties.

- New remote-config-service that handles fetching firebase's remote configs;
- Adds hook for tracking remote config loading state;
- Refactors App's requirements check to a sep. component;
- Removes firebase function calls since they are no longer needed.

- [x] Completes #(CS-3762)

Commit:
Adds new remote-config-service that handles fetching firebase's remote configs;
Adds hook for tracking remote config loading state;
Refactors App's requirements checks to a sep. component;
Deletes firebase function calls, since they are no longer needed.

Commit:
Cant make firebase app's core mock to work, removing for now.

danibonilha commented on package.json:
```diff
@@ -61,8 +61,10 @@
     "@react-native-community/clipboard": "git+https://github.com/gnardini/clipboard#01d1c13",
     "@react-native-community/masked-view": "^0.1.5",
     "@react-native-community/netinfo": "^6.1.0",
+    "@react-native-firebase/analytics": "^14.9.3",
```
is the analytics package required to have remote config working ?

oidoug replied:
Yes, unfortunately. But I found in the docs a secondary analytics package that doesn't require the user to have the AD ID enabled, so it's pretty anonymous.

oidoug replied:
Updated all firebase packages, I looked through the docs, and APIs didn't change.

danibonilha commented on package.json:
```diff
@@ -61,8 +61,10 @@
     "@react-native-firebase/app": "^12.1.0",
     "@react-native-firebase/messaging": "^12.1.0",
+    "@react-native-firebase/remote-config": "^14.9.3",
     "@react-navigation/bottom-tabs": "^5.x",
```
We need to  follow the same version of the other packages, to avoid these warnings

<img width="800" alt="image" src="https://user-images.githubusercontent.com/20520102/168296643-988d14c3-bd9b-44fe-a751-55af687d5f0b.png">

oidoug replied:
Ah sorry, didn't notice the warnings.

oidoug replied:
I updated all firebase packages to the latest. I looked through the docs and APIs are the same.

danibonilha commented on cardstack/src/services/remote-config/remote-config-service.ts:
```diff
@@ -0,0 +1,33 @@
+export enum ConfigKey {
+  requiredMinimumVersion = 'requiredMinimumVersion',
+  maintenanceStatus = 'maintenanceStatus',
+  appReviewFeature = 'appReviewFeature',
```
We can remove this reviewFeature, it's a rainbow leftover

Commit:
Removes AppRequirementsCheck from global comp import;
Removes remote-config service from global service import;
Adds mock to NativeEventEmitter needed for analytics.

danibonilha commented on cardstack/src/services/remote-config/remote-config-service.ts:
```diff
@@ -0,0 +1,33 @@
+
+export const fetchRemoteConfigs = async () => {
+  await remoteConfig().setConfigSettings({
+    minimumFetchIntervalMillis: CACHE_INTERVAL_MILLIS,
```
one hour is still a lot to be refetching something that will hardly change, but also too much for development, but we can set a developer flag to update the data faster, something like an easter egg, let's discuss it later.

oidoug replied:
Default is 12 hrs. I think we can have let it be a few minutes in DEV and a few hours in prod.
Mind that it will not automatically refresh, that's the time the cache is kept as valid.

oidoug replied:
Set to 1min in DEV and 6hrs in prod.

danibonilha commented on cardstack/src/services/remote-config/remote_config_defaults.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "requiredMinimumVersion": "1.1.1",
+  "maintenanceStatus": "{\"active\":false,\"message\":\"Card Wallet is going through scheduled maintenance, please try again later.\"}",
```
This is not so intuitive while updating in the console for example, let's make it a single key, `maintenanceStatusMessage` idk, and if the message exists we render the component OR  we split into two diff keys, being, `maintenanceActive`, and `maintenanceStatusMessage`

oidoug replied:
There's a specific type for JSON properties and the editor validates the input, so it's safe to edit it. But I agree it's more convoluted.

I don't mind splitting into two though, so let's do that then.

danibonilha commented on src/App.js:
```diff
@@ -39,24 +37,22 @@ import {
 import ErrorBoundary from '@cardstack/components/ErrorBoundary/ErrorBoundary';
-import { MinimumVersion } from '@cardstack/components/MinimumVersion';
 import { apolloClient } from '@cardstack/graphql/apollo-client';
+// import { useInitRemoteConfigs } from '@cardstack/hooks';
```
leftover ?

danibonilha commented on cardstack/src/components/AppRequirementsCheck/AppRequirementsCheck.tsx:
```diff
@@ -0,0 +1,46 @@
+    return compareVersions(minVersion, appVersion) > 0;
+  }, [isReady]);
+
+  if (maintenance?.active) {
+    return <MaintenanceMode message={maintenance.message} />;
+  }
+
+  if (forceUpdate) {
+    return <MinimumVersion />;
+  }
```
not hiding the splashScreen on these states, still works as expected ? If so, is the hideSplashScreen really doing something at all?

oidoug replied:
hideSplashScreen was being called regardless of the state, we need to hide the splash if we want anything to display.
So I prefer to have it outside the requirements check.

danibonilha commented on src/App.js:
```diff
@@ -245,53 +242,14 @@ class App extends Component {
 }
 
-const CheckSystemReqs = ({ children }) => {
+const HideSplashScreen = () => {
```
it seems a little odd to have this as a component, if this is to handle the AppRequirements calls, it seems to make more sense to live within the AppReqs component

oidoug replied:
As said above, it's not really part of the AppReqs. It's in a sep. component because our App is a class and can't call our `useHideSplashScreen` hook.

I don't mind doing it differently though.

oidoug replied:
Ah, never mind, I see now what you meant. I'll move the call to be handled by the requirement component again. Strange that it was working.

danibonilha commented on cardstack/src/services/remote-config/remote-config-service.ts:
```diff
@@ -0,0 +1,33 @@
+  remoteConfig().getValue(key).asString();
+
+export const getConfigJSON = (key: ConfigKey) =>
+  JSON.parse(remoteConfig().getValue(key).asString());
```
supernit: if this was a remoteConfig model, we could export these functions within it, and use it like. `remoteConfig.getString` so it's clear that it's related to remoteConfig, cause right now, getConfigJson is too abstract, or maybe, `getRemoteConfigString`

danibonilha commented on cardstack/src/services/remote-config/remote-config-service.ts:
```diff
@@ -0,0 +1,33 @@
+
+  return {
+    fetchedRemotely,
+  };
```
What does it return ? it doesn't seem that the return of this function is being used at all, if the retrieval will always happen within the below functions, this one could be called something like, `loadRemoteConfig` or `setupRemoteConfig` which says that something is happening, but there's not actually values to be used

danibonilha commented on cardstack/src/hooks/remote-configs/useRemoteConfigs.ts:
```diff
@@ -0,0 +1,30 @@
+
+import logger from 'logger';
+
+export const useRemoteConfigs = () => {
```
this name might be misleading as it gives the impression that we can use the remote config values, when it's actually just initializing it

danibonilha commented on cardstack/src/hooks/remote-configs/useRemoteConfigs.ts:
```diff
@@ -0,0 +1,30 @@
+    } catch (error) {
+      logger.sentry("Couldn't fetch remote config", error);
+    } finally {
+      console.log(remoteConfig().getAll());
```
whats the console.log for ?

oidoug replied:
it's a leftover, not needed, I'll remove it.

danibonilha reviewed (commented):
There are somethings that need to be addressed, i'd also have one general suggestion for future tasks like this, to split into smaller chunks, for example, one, setting the library, and creating the models and another for implementing and refactoring.

Commit:
Improves code readability;
Updates all firebase packages to latest;
Fixes splash screen state.

Commit:
Updates remote config cache time to 6hrs

Commit:
Revests change to ios project.

Commit:
Merge branch 'develop' into cs-3762-plan-out-remote-config-needs

danibonilha commented on src/components/settings-menu/SettingsSection.js:
```diff
@@ -70,12 +69,6 @@ export default function SettingsSection({
-    setReviewFeature();
-  }, []);
-
   const onPressReview = useCallback(async () => {
```
We can remove this, as it's not being used anymore

oidoug replied:
Ah, yeah, I merged with dev and that got obsolete.

danibonilha reviewed (approved):
Nice update! Good to go after deleting  the `onPressReview`

Commit:
Adds firebase config for disabling automatic screen reporting in analytics, since it pushes screens with generic names that aren't of much use for us.
We may want to add personalized screen events in the future with logScreenView.

Commit:
Removes last bits from Rainbow review
